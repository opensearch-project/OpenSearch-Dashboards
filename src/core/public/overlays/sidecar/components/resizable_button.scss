/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

.sidecar-resizableButton {
  position: relative;
  flex-shrink: 0;
  z-index: $euiZLevel1;

  &::before,
  &::after {
    content: "";
    display: block;
    position: absolute;
    top: 50%;
    left: 50%;
    background-color: $euiColorDarkestShade;
    transition:
      (width $euiAnimSpeedFast ease,
      height $euiAnimSpeedFast ease,
      transform $euiAnimSpeedFast ease,
      background-color $euiAnimSpeedFast ease);
  }

  &.resizableButton--horizontal {
    cursor: col-resize;
    width: $euiSize;
    margin-left: calc(-1 * $euiSize / 2);
    margin-right: calc(-1 * $euiSize / 2);

    &::before,
    &::after {
      width: 1px;
      height: $euiSizeM;
    }

    &::before {
      transform: translate(-2px, -50%);
    }

    &::after {
      transform: translate(1px, -50%);
    }
  }

  &.resizableButton--vertical {
    cursor: row-resize;
    height: $euiSize;
    margin-top: calc(-1 * $euiSize / 2);
    margin-bottom: calc(-1 * $euiSize / 2);

    &::before,
    &::after {
      width: $euiSizeM;
      height: 1px;
    }

    &::before {
      transform: translate(-50%, -2px);
    }

    &::after {
      transform: translate(-50%, 1px);
    }
  }

  // Lighten the "grab" icon on :hover
  &:hover:not(:disabled) {
    &::before,
    &::after {
      background-color: $euiColorMediumShade;
      transition-delay: $euiAnimSpeedFast;
    }
  }

  // Add a transparent background to the container and emphasis the "grab" icon with primary color on :focus
  &:focus:not(:disabled) {
    background-color: transparentize($euiColorPrimary, 0.9);

    &::before,
    &::after {
      background-color: $euiColorPrimary;

      // Overrides default transition so that "grab" icon background-color doesn't animate
      transition:
        (width $euiAnimSpeedFast ease,
        height $euiAnimSpeedFast ease,
        transform $euiAnimSpeedFast ease);
      transition-delay: calc($euiAnimSpeedFast / 2);
    }
  }

  // Morph the "grab" icon into a fluid 2px straight line on :hover and :focus
  &:hover:not(:disabled),
  &:focus:not(:disabled) {
    &.resizableButton--horizontal {
      &::before,
      &::after {
        height: 100%;
      }

      &::before {
        transform: translate(-1px, -50%);
      }

      &::after {
        transform: translate(0, -50%);
      }
    }

    &.resizableButton--vertical {
      &::before,
      &::after {
        width: 100%;
      }

      &::before {
        transform: translate(-50%, -1px);
      }

      &::after {
        transform: translate(-50%, 0);
      }
    }
  }

  &:disabled {
    display: none !important; // sass-lint:disable-line no-important
  }
}
