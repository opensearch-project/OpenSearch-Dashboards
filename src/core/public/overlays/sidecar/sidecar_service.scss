/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

.osdSidecarFlyout {
  box-shadow: initial;
  position: fixed;
  z-index: 1000;
  background: white;
  display: flex;

  &.osdSidecarFlyout--dockedRight {
    top: 0;
    height: 100vh;
    right: 0;
    flex-direction: row;
  }

  &.osdSidecarFlyout--dockedLeft {
    top: 0;
    height: 100vh;
    left: 0;
    flex-direction: row-reverse;
  }

  &.osdSidecarFlyout--dockedTakeover {
    width: 100vw !important;
    bottom: 0;
    flex-direction: column;
  }

  &.osdSidecarFlyout--hide {
    display: none;
  }

  .osdSidecarMountWrapper {
    width: 100%;
    height: 100%;
  }
}
