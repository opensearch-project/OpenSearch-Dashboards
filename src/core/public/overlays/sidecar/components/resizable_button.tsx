/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useRef } from 'react';
import classNames from 'classnames';
import './resizable_button.scss';
import { getPosition } from '../helper';
import { ISidecarConfig, SIDECAR_DOCKED_MODE } from '../sidecar_service';

interface Props {
  onResize: (size: number) => void;
  flyoutSize: number;
  dockedMode: ISidecarConfig['dockedMode'] | undefined;
}

export const MIN_SIDECAR_SIZE = 350;

export const ResizableButton = ({ dockedMode, onResize, flyoutSize }: Props) => {
  const isHorizontal = dockedMode !== SIDECAR_DOCKED_MODE.TAKEOVER;

  const classes = classNames('sidecar-resizableButton', {
    'resizableButton--vertical': !isHorizontal,
    'resizableButton--horizontal': isHorizontal,
  });

  const initialMouseXorY = useRef(0);
  const initialFlyoutSize = useRef(flyoutSize);
  const setFocus = (e: React.MouseEvent<HTMLButtonElement>) => e.currentTarget.focus();

  const onMouseDown = useCallback(
    (event: React.MouseEvent | React.TouchEvent) => {
      const onMouseUp = () => {
        initialMouseXorY.current = 0;
        window.removeEventListener('mousemove', onMouseMove);
        window.removeEventListener('mouseup', onMouseUp);
        window.removeEventListener('touchmove', onMouseMove);
        window.removeEventListener('touchend', onMouseUp);
      };
      const onMouseMove = (e: MouseEvent | TouchEvent) => {
        let offset;
        if (dockedMode === SIDECAR_DOCKED_MODE.LEFT) {
          offset = getPosition(e, isHorizontal) - initialMouseXorY.current;
        } else {
          offset = initialMouseXorY.current - getPosition(e, isHorizontal);
        }
        const newFlyoutSize = initialFlyoutSize.current + offset;

        onResize(Math.max(newFlyoutSize, MIN_SIDECAR_SIZE));
      };
      window.addEventListener('mousemove', onMouseMove);
      window.addEventListener('mouseup', onMouseUp);
      window.addEventListener('touchmove', onMouseMove);
      window.addEventListener('touchend', onMouseUp);
      initialMouseXorY.current = getPosition(event, isHorizontal);
      initialFlyoutSize.current = flyoutSize;
    },
    [isHorizontal, flyoutSize, dockedMode, onResize]
  );

  return (
    <button
      className={classes}
      data-test-subj="resizableButton"
      type="button"
      onClick={setFocus}
      onMouseDown={onMouseDown}
      onTouchStart={onMouseDown}
    />
  );
};
