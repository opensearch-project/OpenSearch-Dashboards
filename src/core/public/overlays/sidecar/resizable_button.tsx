/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useRef } from 'react';
import classNames from 'classnames';
import './rsizable_button.scss';
import { getPosition } from './helper';
import { ISidecarConfig, SIDECAR_DOCKED_MODE } from './sidecar_service';

interface Props {
  isHorizontal: boolean;
  onResize: (size: number) => void;
  flyoutSize: number;
  dockedMode: ISidecarConfig['dockedMode'] | undefined;
  minSize?: number;
}

const MIN_SIDECAR_SIZE = 200;

export const ResizableButton = ({
  dockedMode,
  onResize,
  flyoutSize,
  minSize = MIN_SIDECAR_SIZE,
}: Props) => {
  const isHorizontal = dockedMode !== SIDECAR_DOCKED_MODE.TAKEOVER;

  const classes = classNames('resizableButton', {
    'resizableButton--vertical': !isHorizontal,
    'resizableButton--horizontal': isHorizontal,
  });

  const initialMouseXorY = useRef(0);
  const initialFlyoutSize = useRef(flyoutSize);
  const setFocus = (e: React.MouseEvent<HTMLButtonElement>) => e.currentTarget.focus();

  const onMouseDown = useCallback(
    (event: React.MouseEvent | React.TouchEvent) => {
      const onMouseUp = () => {
        initialMouseXorY.current = 0;
        window.removeEventListener('mousemove', onMouseMove);
        window.removeEventListener('mouseup', onMouseUp);
        window.removeEventListener('touchmove', onMouseMove);
        window.removeEventListener('touchend', onMouseUp);
      };
      const onMouseMove = (e: MouseEvent | TouchEvent) => {
        let offset;
        if (dockedMode === 'left') {
          offset = getPosition(e, isHorizontal) - initialMouseXorY.current;
        } else {
          offset = initialMouseXorY.current - getPosition(e, isHorizontal);
        }
        const newFlyoutSize = initialFlyoutSize.current + offset;

        onResize(Math.max(newFlyoutSize, minSize));
      };
      window.addEventListener('mousemove', onMouseMove);
      window.addEventListener('mouseup', onMouseUp);
      window.addEventListener('touchmove', onMouseMove);
      window.addEventListener('touchend', onMouseUp);
      initialMouseXorY.current = getPosition(event, isHorizontal);
      initialFlyoutSize.current = flyoutSize;
    },
    [isHorizontal, flyoutSize, dockedMode, minSize, onResize]
  );

  return (
    <button
      className={classes}
      data-test-subj="resizableButton"
      type="button"
      onClick={setFocus}
      onMouseDown={onMouseDown}
      onTouchStart={onMouseDown}
    />
  );
};
