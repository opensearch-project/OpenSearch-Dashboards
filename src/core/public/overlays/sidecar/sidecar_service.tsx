/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
/* eslint-disable max-classes-per-file */

import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { BehaviorSubject, Observable, ReplaySubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { I18nStart } from '../../i18n';
import { MountPoint } from '../../types';
import { OverlayRef } from '../types';
import { Sidecar } from './components/sidecar';
/**
 * A SidecarRef is a reference to an opened sidecar panel. It offers methods to
 * close the sidecar panel again. If you open a sidecar panel you should make
 * sure you call `close()` when it should be closed.
 * Since a sidecar could also be closed by a user or from another sidecar being
 * opened, you must bind to the `onClose` Promise on the Sidecar instance.
 * The Promise will resolve whenever the sidecar was closed at which point you
 * should discard the SidecarRef.
 *
 * @public
 */
class SidecarRef implements OverlayRef {
  /**
   * An Promise that will resolve once this sidecar is closed.
   *
   * Sidecars can close from user interaction, calling `close()` on the sidecar
   * reference or another call to `open()` replacing your sidecar.
   */
  public readonly onClose: Promise<void>;

  private closeSubject = new Subject<void>();

  constructor() {
    this.onClose = this.closeSubject.toPromise();
  }

  /**
   * Closes the referenced sidecar if it's still open which in turn will
   * resolve the `onClose` Promise. If the sidecar had already been
   * closed this method does nothing.
   */
  public close(): Promise<void> {
    if (!this.closeSubject.closed) {
      this.closeSubject.next();
      this.closeSubject.complete();
    }
    return this.onClose;
  }
}

/**
 * APIs to open and manage sidecar.
 *
 * @public
 */
export interface OverlaySidecarStart {
  /**
   * Opens a sidecar panel with the given mount point inside. You can use
   * `close()` on the returned SidecarRef to close the sidecar.
   *
   * @param mount {@link MountPoint} - Mounts the children inside a sidecar panel
   * @param options {@link OverlaySidecarOpenOptions} - options for the sidecar
   * @return {@link SidecarRef} A reference to the opened sidecar panel.
   */
  open(mount: MountPoint, options: OverlaySidecarOpenOptions): SidecarRef;

  /**
   * Override the default support config
   * @param config The updated support config
   */
  setSidecarConfig(config: Partial<ISidecarConfig>): void;

  /**
   * Get an observable of the current locked state of the nav drawer.
   */
  getSidecarConfig$(): Observable<ISidecarConfig | undefined>;

  /**
   * Hide side car.
   */
  hide: () => void;

  /**
   * Show side car.
   */
  show: () => void;
}

/**
 * @public
 */
export interface OverlaySidecarOpenOptions {
  className?: string;
  'data-test-subj'?: string;
  config: ISidecarConfig;
}

interface StartDeps {
  i18n: I18nStart;
  targetDomElement: Element | null;
}

export enum SIDECAR_DOCKED_MODE {
  LEFT = 'left',
  RIGHT = 'right',
  TAKEOVER = 'takeover',
}
export interface ISidecarConfig {
  // takeover mode will docked to bottom
  dockedMode: SIDECAR_DOCKED_MODE;
  paddingSize: number;
  isHidden?: boolean;
}

/** @internal */
export class SidecarService {
  private activeSidecar: SidecarRef | null = null;
  private targetDomElement: Element | null = null;
  private readonly stop$ = new ReplaySubject(1);

  public start({ i18n, targetDomElement }: StartDeps): OverlaySidecarStart {
    this.targetDomElement = targetDomElement;
    const sidecarConfig$ = new BehaviorSubject<ISidecarConfig | undefined>(undefined);
    const getSidecarConfig$ = () => sidecarConfig$.pipe(takeUntil(this.stop$));

    const setSidecarConfig = (config: Partial<ISidecarConfig>) => {
      if (
        sidecarConfig$.value ||
        // init
        (!sidecarConfig$.value && config.dockedMode && config.paddingSize)
      ) {
        sidecarConfig$.next({ ...sidecarConfig$.value, ...config } as ISidecarConfig);
      }
    };

    return {
      getSidecarConfig$,
      setSidecarConfig,
      open: (mount: MountPoint, options: OverlaySidecarOpenOptions): SidecarRef => {
        // If there is an active flyout session close it before opening a new one.
        if (this.activeSidecar) {
          setSidecarConfig({ paddingSize: 0 });
          this.activeSidecar.close();
          this.cleanupDom();
        }

        const sidecar = new SidecarRef();
        setSidecarConfig(options.config);

        // If a sidecar gets closed through it's SidecarRef, remove it from the dom
        sidecar.onClose.then(() => {
          if (this.activeSidecar === sidecar) {
            this.cleanupDom();
          }
        });

        this.activeSidecar = sidecar;

        render(
          <Sidecar
            sidecarConfig$={sidecarConfig$}
            setSidecarConfig={setSidecarConfig}
            options={options}
            i18n={i18n}
            mount={mount}
          />,
          this.targetDomElement
        );

        return sidecar;
      },
      hide: () => {
        setSidecarConfig({ isHidden: true });
      },
      show: () => {
        setSidecarConfig({ isHidden: false });
      },
    };
  }

  /**
   * When using React.render to re-render content into a target DOM element, it replaces
   * the content without invoking unmountComponentAtNode on any components within the target
   * or their children. To avoid potential subtle bugs in child components that rely on
   * unmounting for cleanup behavior, this function ensures proper cleanup of the DOM.
   */
  private cleanupDom(): void {
    if (this.targetDomElement != null) {
      unmountComponentAtNode(this.targetDomElement);
      this.targetDomElement.innerHTML = '';
    }
    this.activeSidecar = null;
  }
}
