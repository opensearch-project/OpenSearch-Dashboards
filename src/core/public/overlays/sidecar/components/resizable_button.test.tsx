/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';

import { ResizableButton, MIN_SIDECAR_SIZE } from './resizable_button';
import { shallow } from 'enzyme';
import { SIDECAR_DOCKED_MODE } from '../sidecar_service';

const storeWindowEvents = () => {
  const map: Record<string, Function> = {};
  window.addEventListener = jest.fn().mockImplementation((event: string, cb) => {
    map[event] = cb;
  });
  return map;
};

const DEFAULT_FLYOUT_SIZE = 460;
const props = {
  dockedMode: SIDECAR_DOCKED_MODE.RIGHT,
  onResize: jest.fn(),
  flyoutSize: DEFAULT_FLYOUT_SIZE,
};

test('is rendered', () => {
  const component = shallow(<ResizableButton {...props} />);
  expect(component).toMatchSnapshot();
});

test('it should be horizontal when docked mode is right', () => {
  const component = shallow(<ResizableButton {...props} />);
  expect(component.hasClass('resizableButton--horizontal')).toBe(true);
  expect(component.hasClass('resizableButton--vertical')).toBe(false);
  expect(component).toMatchSnapshot();
});

test('it should be vertical when docked mode is takeover', () => {
  const newProps = { ...props, dockedMode: SIDECAR_DOCKED_MODE.TAKEOVER };
  const component = shallow(<ResizableButton {...newProps} />);
  expect(component.hasClass('resizableButton--vertical')).toBe(true);
  expect(component.hasClass('resizableButton--horizontal')).toBe(false);
  expect(component).toMatchSnapshot();
});

test('it should emit onResize with new flyout size when docked right and drag and horizontal', () => {
  const windowEvents = storeWindowEvents();

  const onResize = jest.fn();
  const newProps = { ...props, onResize };
  const component = shallow(<ResizableButton {...newProps} />);
  const resizer = component.find(`[data-test-subj~="resizableButton"]`).first();
  expect(onResize).not.toHaveBeenCalled();
  resizer.simulate('mousedown', { clientX: 0, pageX: 0, pageY: 0 });
  windowEvents?.mousemove({ clientX: -1000, pageX: 0, pageY: 0 });
  windowEvents?.mouseup();
  const newSize = 1000 + DEFAULT_FLYOUT_SIZE;
  expect(onResize).toHaveBeenCalledWith(newSize);
  expect(component).toMatchSnapshot();
});

test('it should emit onResize with new flyout size when docked left and drag and horizontal', () => {
  const windowEvents = storeWindowEvents();

  const onResize = jest.fn();
  const newProps = { ...props, onResize, dockedMode: SIDECAR_DOCKED_MODE.LEFT };
  const component = shallow(<ResizableButton {...newProps} />);
  const resizer = component.find(`[data-test-subj~="resizableButton"]`).first();
  expect(onResize).not.toHaveBeenCalled();
  resizer.simulate('mousedown', { clientX: 0, pageX: 0, pageY: 0 });
  windowEvents?.mousemove({ clientX: 1000, pageX: 0, pageY: 0 });
  windowEvents?.mouseup();
  const newSize = 1000 + DEFAULT_FLYOUT_SIZE;
  expect(onResize).toHaveBeenCalledWith(newSize);
  expect(component).toMatchSnapshot();
});

test('it should emit onResize with new flyout size when drag and vertical', () => {
  const windowEvents = storeWindowEvents();

  const onResize = jest.fn();
  const newProps = { ...props, onResize, dockedMode: SIDECAR_DOCKED_MODE.TAKEOVER };
  const component = shallow(<ResizableButton {...newProps} />);
  const resizer = component.find(`[data-test-subj~="resizableButton"]`).first();
  expect(onResize).not.toHaveBeenCalled();
  resizer.simulate('mousedown', { clientY: 0, pageX: 0, pageY: 0 });
  windowEvents?.mousemove({ clientY: -1000, pageX: 0, pageY: 0 });
  windowEvents?.mouseup();
  const newSize = 1000 + DEFAULT_FLYOUT_SIZE;
  expect(onResize).toHaveBeenCalledWith(newSize);
  expect(component).toMatchSnapshot();
});

test('it should emit onResize with min size when drag if new size is below the minimum', () => {
  const windowEvents = storeWindowEvents();
  const onResize = jest.fn();
  const newProps = { ...props, onResize };
  const component = shallow(<ResizableButton {...newProps} />);
  const resizer = component.find(`[data-test-subj~="resizableButton"]`).first();
  expect(onResize).not.toHaveBeenCalled();
  resizer.simulate('mousedown', { clientX: 0, pageX: 0, pageY: 0 });
  windowEvents?.mousemove({ clientX: 1000, pageX: 0, pageY: 0 });
  windowEvents?.mouseup();
  const newSize = MIN_SIDECAR_SIZE;
  expect(onResize).toHaveBeenCalledWith(newSize);
  expect(component).toMatchSnapshot();
});
