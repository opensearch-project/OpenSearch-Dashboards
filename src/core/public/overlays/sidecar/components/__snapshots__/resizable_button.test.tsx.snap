// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is rendered 1`] = `
<button
  className="sidecar-resizableButton resizableButton--horizontal"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;

exports[`it should be horizontal when docked mode is right 1`] = `
<button
  className="sidecar-resizableButton resizableButton--horizontal"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;

exports[`it should be vertical when docked mode is takeover 1`] = `
<button
  className="sidecar-resizableButton resizableButton--vertical"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;

exports[`it should emit onResize with min size when drag if new size is below the minimum 1`] = `
<button
  className="sidecar-resizableButton resizableButton--horizontal"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;

exports[`it should emit onResize with new flyout size when docked left and drag and horizontal 1`] = `
<button
  className="sidecar-resizableButton resizableButton--horizontal"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;

exports[`it should emit onResize with new flyout size when docked right and drag and horizontal 1`] = `
<button
  className="sidecar-resizableButton resizableButton--horizontal"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;

exports[`it should emit onResize with new flyout size when drag and vertical 1`] = `
<button
  className="sidecar-resizableButton resizableButton--vertical"
  data-test-subj="resizableButton"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  type="button"
/>
`;
